### PR TITLE
Add package: Jimp

### PR DESCRIPTION
**Pull request by Jack-Works**
Signed-off-by: Jack Works <zjwpeter@gmail.com>
https://github.com/oliver-moran/jimp
- [OK] The package does not provide its own types, and you can not add them.

### CODE REVIEW

File: jimp/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	type PresetFont = string;
+	type FileMINE = string;
+	type AlignMode = number;
+	type FilterType = number;
+	type Font = any | PresetFont;
+
+
+	class JimpImage {
+		constructor(width: number, height: number, callback?: Callback<JimpImage>);
+		constructor(width: number, height: number, initialColor: number, callback?: Callback<JimpImage>);
```
Comment (zhengbli) [inline jimp/index.d.ts]: Combine the two, use
`constructor(width: number, height: number, initialColor?: number, callback?: Callback<JimpImage>)`

File: jimp/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
+// Type definitions for jimp 0.2
+// Project: https://github.com/oliver-moran/jimp#readme
+// Definitions by: Jack Works <https://github.com/Jack-Works>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/// <reference types="node" />
+
```
Comment (zhengbli) [inline jimp/index.d.ts]: Per the guideline, can you use either 4 spaces or all tabs for indentation instead? Thanks

File: jimp/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+		rgba(is: boolean): JimpImage;
+		filterType(type: FilterType): JimpImage;
+		deflateLevel(level: number): JimpImage;
+
+		/* Operation */;
+		clone(): JimpImage;
+		print(font: Font, x: number, y: number, text: string, width?: number): JimpImage;
+
+		/* Adv */;
+		color(params: Array<{apply: 'lighten' | 'brighten' | 'darken' |
```
Comment (zhengbli) [inline jimp/index.d.ts]: This formatting looks wierd. Can you put the parameters in a new line if it is too long?

File: jimp/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+									'desaturate' | 'saturate' | 'greyscale' |
+									'spin' | 'hue' | 'tint' | 'shade' | 'red' | 'green' | 'blue', params: [number]
+			} | {apply: 'mix', params: [string, number];
+		} |{apply: 'xor', params: [string]}>): JimpImage;
+		;
```
Comment (zhengbli) [inline jimp/index.d.ts]: Remove extra `;`

File: jimp/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+		contain(width: number, height: number, callback?: Callback<JimpImage>): JimpImage;
+		contain(width: number, height: number, alignBitsOrMode: number | ResizeMode, callback?: Callback<JimpImage>): JimpImage;
+		contain(width: number, height: number, alignBits: number, mode: ResizeMode, callback?: Callback<JimpImage>): JimpImage;
+
+		cover(width: number, height: number, callback?: Callback<JimpImage>): JimpImage;
+		cover(width: number, height: number, alignBitsOrmode: ResizeMode | number, callback?: Callback<JimpImage>): JimpImage;
+		cover(width: number, height: number, alignBits: number, mode?: ResizeMode, callback?: Callback<JimpImage>): JimpImage;
+
+		resize(width: number, height: number, callback?: Callback<JimpImage>): JimpImage;
+		resize(width: number, height: number, mode?: ResizeMode, callback?: Callback<JimpImage>): JimpImage;
```
Comment (zhengbli) [inline jimp/index.d.ts]: The first signature is covered by the second one, so no longer needed.

File: jimp/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+		scale(factor: any, callback?: Callback<JimpImage>): JimpImage;
+		scale(factor: any, mode: ResizeMode, callback?: Callback<JimpImage>): JimpImage;
```
Comment (zhengbli) [inline jimp/index.d.ts]: Same thing. move `mode` optional and remove the first signature

File: jimp/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+		scaleToFit(width: number, height: number, callback?: Callback<JimpImage>): JimpImage;
+		scaleToFit(width: number, height: number, mode?: ResizeMode, callback?: Callback<JimpImage>): JimpImage;
```
Comment (zhengbli) [inline jimp/index.d.ts]: and here

File: jimp/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+		/* Crop */;
```
Comment (zhengbli) [inline jimp/index.d.ts]: no semicolon after comments

File: jimp/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+		autocrop(callback?: Callback<JimpImage>): JimpImage;
+		crop(x: number, y: number, width: number, height: number, callback?: Callback<JimpImage>): JimpImage;
+
+		/* Composing */;
+		blit(scr: string, x: number, y: number, callback?: Callback<JimpImage>): JimpImage;
+		blit(scr: string, x: number, y: number, srcx?: number, srcy?: number, callback?: Callback<JimpImage>): JimpImage;
+		blit(scr: string, x: number, y: number, srcx?: number, srcy?: number, srcw?: number, srch?: number, callback?: Callback<JimpImage>): JimpImage;
```
Comment (zhengbli) [inline jimp/index.d.ts]: The last one covers all three.

File: jimp/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+		composite(scr: string, x: number, y: number, callback?: Callback<JimpImage>): JimpImage;
+		mask(scr: string, x: number, y: number, callback?: Callback<JimpImage>): JimpImage;
+
+		/* Flip and rotate */;
+		flip(horz: any, vert: any, callback?: Callback<JimpImage>): JimpImage;
+		mirror(horz: any, vert: any, callback?: Callback<JimpImage>): JimpImage;
+		rotate(deg: number, callback?: Callback<JimpImage>): JimpImage;
+		rotate(deg: number, mode: any, callback?: Callback<JimpImage>): JimpImage;
```
Comment (zhengbli) [inline jimp/index.d.ts]: same here, please also revise the rest similar cases.

File: jimp/index.d.ts [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+
+		/* Blurs */;
+		gaussian(pixel: number, callback?: Callback<JimpImage>): JimpImage;
+		blur(pixel: number, callback?: Callback<JimpImage>): JimpImage;
+
+		/* Effects */;
+		posterize(level: number, callback?: Callback<JimpImage>): JimpImage;
+		sepia(callback?: Callback<JimpImage>): JimpImage;
+	}
+	class JIMP extends JimpImage {;
```
Comment (zhengbli) [inline jimp/index.d.ts]: remove `;`

File: jimp/index.d.ts [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+		static RESIZE_HERMITE: ResizeMode;
+		static RESIZE_BEZIER: ResizeMode;
+
+		static FONT_SANS_8_BLACK: PresetFont;
+		static FONT_SANS_16_BLACK: PresetFont;
+		static FONT_SANS_32_BLACK: PresetFont;
+		static FONT_SANS_64_BLACK: PresetFont;
+		static FONT_SANS_128_BLACK: PresetFont;
+
+		static FONT_SANS_8_WHITE: PresetFont;
```
Comment (zhengbli) [inline jimp/index.d.ts]: if this is a constant, make it `readonly`

**Review by zhengbli — CHANGES_REQUESTED** [review]

File: jimp/index.d.ts (hunk 1 above)
Comment (Jack-Works) [inline jimp/index.d.ts]: Um... Combine this two will make code does not type check
`new Jimp(256, 256, function (err, image) {});`

File: jimp/index.d.ts (hunk 2 above)
Comment (Jack-Works) [inline jimp/index.d.ts]: I can't find where I use space for indentaton in this file...

File: jimp/index.d.ts (hunk 5 above)
Comment (Jack-Works) [inline jimp/index.d.ts]: Same, this will make `image.resize(x, y, function(){})` does not type check

File: jimp/index.d.ts (hunk 6 above)
Comment (Jack-Works) [inline jimp/index.d.ts]: Same, will make `image.scale(0.5, function (){})` does not type check

File: jimp/index.d.ts (hunk 7 above)
Comment (Jack-Works) [inline jimp/index.d.ts]: Same thing

File: jimp/index.d.ts (hunk 9 above)
Comment (Jack-Works) [inline jimp/index.d.ts]: No, will break type check

**Review by zhengbli — APPROVED** [review]

**Comment by zhengbli** [comment]
looks good, I'll merge it after the CI errors are fixed. Thank you!

File: jimp/index.d.ts [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+			} | {
+				apply: 'mix',
+				params: [string, number]
+			} | {
+				apply: 'xor',
+				params: [string]
+			}>): JimpImage;
+
+		convolution(matrix: number[][]): JimpImage;
+		scan(x: number, y: number, width: number, height: number, callback?: (this: JimpImage, x?: number, y?: number, idx?: number) => void): JimpImage;
```
Comment (hinell) [inline jimp/index.d.ts]: Optional callback parameters [are senseless](https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/declaration%20files/Do's%20and%20Don'ts.md#optional-parameters-in-callbacks).

**Review by hinell — CHANGES_REQUESTED** [review]

**Review by hinell — APPROVED** [review]

**Pull request merged**